BOLT 7: Onion message support (features 38/39) (#759)
lightning/bolts: 04-onion-routing.md

    2. data:
        * [`...*byte`:`enctlv`]
    1. type: 8 (`reply_path`)
    2. data:
        * [`point`:`blinding`]
        * [`...*onionmsg_path`:`path`]
    1. type: 10 (`blinding`)
    2. data:
        * [`point`:`blinding`]

1. tlvs: `encmsg_tlvs`
2. types:
    1. type: 2 (`next_short_channel_id`)
    2. data:
        * [`short_channel_id`:`short_channel_id`]
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]

1. subtype: `onionmsg_path`
2. data:
    * [`point`:`node_id`]
    * [`u16`:`enclen`]
    * [`enclen*byte`:`enctlv`]

#### Requirements

The writer:
- For the non-final nodes' `onionmsg_payload`:
  - MUST include exactly one of `next_short_channel_id`, `next_node_id`
    or `enctlv` indicating the next node.
- For the the final node's `onionmsg_payload`:
  - if the final node is permitted to reply:
	- MUST set `reply_path` `blinding` to the initial blinding factor for the `next_node_id`
    - For the first `reply_path` `path`:
	  - MUST set `node_id` to the first node in the reply path.
	- For the remaining `reply_path` `path`:
	  - MUST set `node_id` to the blinded node id to encrypt the onion hop for.
    - Within `reply_path` `path`:

>       - MUST set `enctlv` to the ChaCha20 encryption of a valid `encmsg_tlvs` containing exactly

Reviewer: ChaCha20-Poly1305?

Author: Yes, with all-zero nonce.  I fixed to xref into the (not yet included!) blinded path spec.

Onion messages have an onion with an alternate `hop_payload`
format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
is no legacy format, thus a `bigsize` of 0 means no payload.

1. tlvs: `onionmsg_payload`
2. types:
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]
    1. type: 6 (`next_short_channel_id`)
    2. data:
        * [`short_channel_id`:`short_channel_id`]
    1. type: 8 (`reply_path`)
    2. data:
        * [`point`:`blinding`]
        * [`...*onionmsg_path`:`path`]
    1. type: 10 (`enctlv`)
    2. data:
        * [`...*byte`:`enctlv`]
    1. type: 12 (`blinding`)
    2. data:
        * [`point`:`blinding`]

1. tlvs: `encmsg_tlvs`
2. types:
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]
    1. type: 6 (`next_short_channel_id`)
    2. data:
        * [`short_channel_id`:`short_channel_id`]

1. subtype: `onionmsg_path`
2. data:
    * [`point`:`node_id`]
    * [`u16`:`enclen`]
    * [`enclen*byte`:`enctlv`]

#### Requirements

The writer:
- For the non-final nodes' `onionmsg_payload`:
  - MUST include exactly one of `next_short_channel_id`, `next_node_id`
    or `enctlv` indicating the next node.

> - For the the final node's `onionmsg_payload`:

Reviewer 2: s/the the/the/

invoice timeout), however if it were to fulfill only some of them,
intermediary nodes could simply claim the remaining ones.

### Onion Messages

> Onion messages have an onion with an alternate `hop_payload`

Reviewer 3: Please drop the reference to `hop_payload` and just specify the whole message. It takes a ton of reading to maybe kinda understand how the message is laid out on the wire, instead of being able to look at something.

Reviewer 4: Having recently scrolled around up and down to figure out the message format, strongly agree here.

Author: Yeah, this gets simpler when we remove the legacy format altogether.  But I've rewritten it to duplicate the full description of the onion (with references back, since you don't need to reimplement it but should share code!).

> format: a `bigsize` followed by a `onionmsg_payload`.  Note that there

Reviewer 3: Please, no bigsize, just drop it entirely and use the 2 byte length in the `onion_message` struct.

Author: This matches the modern onion format in htlc onions, which means no fresh parser to write.

Reviewer 3: Huh, its a fresh parser either way because a completely different TLV namespace? I guess your length read is at a different layer from the fields read, then?

Author: Yes, we have wrap/unwrap routines, which just needed to be rejigged to allow a size param (rather than 1300 hardcoded).  Was trying to KISS...

is no legacy format, thus a `bigsize` of 0 means no payload.

An intermediate node expects an `enctlv` which it can decrypt using
the `blinding` which it is handed along with the onion message.

1. `tlv_stream`: `onionmsg_payload`
2. types:
    1. type: 8 (`reply_path`)
    2. data:
        * [`point`:`first_node_id`]
        * [`point`:`blinding`]
        * [`...*onionmsg_path`:`path`]
    1. type: 10 (`enctlv`)
    2. data:
        * [`...*byte`:`enctlv`]

1. `tlv_stream`: `encmsg_tlvs`
2. types:
    1. type: 1 (`padding`)
    2. data:
        * [`...*byte`:`pad`]
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]
    1. type: 12 (`next_blinding`)
    2. data:
        * [`point`:`blinding`]

1. subtype: `onionmsg_path`
2. data:
    * [`point`:`node_id`]
    * [`u16`:`enclen`]

>     * [`enclen*byte`:`enctlv`]

Reviewer 3: It is not specified at all what this is, nor does route blinding tell you what it is, as far as I can tell. The route blinding doc does not either, or at least if it does you don't specify where to look for it.

Reviewer: Actually now that I realize that onion messages uses route blinding on the way forward as well, it's not encrypted by the recipient but by the sender (but on the reply path it will be encrypted by the recipient), so my name sucks...

I cannot do better than `encrypted_data` then, wdyt?

#### Requirements

The writer:
- For the non-final nodes' `onionmsg_payload`:
  - MUST set `enctlv` to a valid `encmsg_tlvs` containing `next_node_id`.
  - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].

>     `ChaChaPoly-1305` encryption using an all-zero nonce).

Reviewer 3: Its really kinda awkward to use ChaCha+HMAC-SHA256 for the outer onion and ChaCha+Poly1305 for the inner one. I guess you decided that saving the 16 bytes per hop is that important? But if so, why not just also use ChaChaPoly for the outer onion as well?

Author: Again, this is fixed already unless we want to rewrite all onion code.  The inner one is better chachapoly1305, because that we can control!

Reviewer 3: Heh, on our end almost everything is abstracted so that the decrypt is 5 lines of code, but I can still trivially replace the MAC call cause which method is called is not abstracted. I get that your code isn't that way, but man it makes for a really awkward spec. I guess there's no risk for the creator to do any collisions and saving 16 bytes is worth it, but it may be worth it on your end to see if you can tweak things a bit to make reuse 95% of the onion decrypt code without being boxed into ChaCha-SHA256.

Reviewer: As discussed during the last spec meeting, roasbeef provided a pretty good answer: crypto libraries usually don't let you use poly1305 without chacha20, which is why hmac-sha256 was initially chosen.

invoice timeout), however if it were to fulfill only some of them,
intermediary nodes could simply claim the remaining ones.

### Onion Messages

Onion messages have an onion with an alternate `hop_payload`
format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
is no legacy format, thus a `bigsize` of 0 means no payload.

An intermediate node expects an `enctlv` which it can decrypt using
the `blinding` which it is handed along with the onion message.

> 1. `tlv_stream`: `onionmsg_payload`

Reviewer 3: I think I mentioned this off GitHub, but I really think there needs to be a separate TLV stream for the actual message content. Shoving it in the same one is cool, but in practice if we have multiple message types we're gonna want to pass a whole TLV stream to different parsers, not have each message type hook the same TLV stream parsing logic.

Author: You mean we have a `payload` blob of bytes at this level, and YA tlv inside it?  Possible, but I don't think demux is going to get any prettier?

Reviewer 3: That's one way to accomplish it, yes. I don't really understand how to demux without some second-level thing here - we have a tlv stream that contains fields that the onion message decoder/processor needs to handle, and fields that the onion-message-type handler needs to process. Presumably those need to be two separate things because there could be different types of onion messages. Arguably there needs to be, per type of onion message, a different, single TLV record in the payload here that declares the type and has all the type-specific data (eg type 1 would be "BOLT 12 offer request" and type 3 would be "BOLT 12 offer response", and....etc).

Author: Perhaps we just say "64 and above"?  It is much nice to use one number per message type, IMHO, rather than one per feature.

Reviewer: SGTM, let's keep it like this for now and we can have this discussion when we review features that actually add tlvs for the final node.

2. types:
    1. type: 2 (`reply_path`)
    2. data:
        * [`point`:`first_node_id`]
        * [`point`:`blinding`]
        * [`...*onionmsg_path`:`path`]
    1. type: 10 (`enctlv`)
    2. data:
        * [`...*byte`:`enctlv`]

1. `tlv_stream`: `encmsg_tlvs`
2. types:
    1. type: 1 (`padding`)
    2. data:
        * [`...*byte`:`pad`]
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]
    1. type: 12 (`next_blinding`)
    2. data:
        * [`point`:`blinding`]

>     1. type: 14 (`self_id`)

Reviewer 5: I don't think `self_id` is an informative name. I think it should be `path_id` because it identifies the reply path.

Author: Sure.

    2. data:

>         * [`...*byte`:`data`]

Reviewer 5: Isn't the `self_id` intended for the final recipient? I thought that `enctlv` was only for intermediate nodes and `reply_path` only for the final recipient. If that's the case, then `self_id` should be in the `onionmsg_payload` next to `reply_path` and not nested inside a `enctlv` that is not needed for the final node.

Author: enctlv is optional for the final recipient, but in practice you need it, to contain *some* identifying information (proposed here as `self_id`) as soon as you start using onion messages.  Otherwise you need to remember the blinding factor for every single path you've ever sent out, to avoid responding inappropriately.

Reviewer 5: I understood why you want the `self_id`, I just thought it would be simpler to put it directly in the `onionmsg_payload`, not in the `enctlv`. I've now realized that it needs to be in the `enctlv` from the `reply_path` so that the `self_id` can't be modified by the person who replies.

    - For every `reply_path` `path`:
      - MUST set `node_id` to the blinded node id to encrypt the onion hop for.
      - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
        `ChaChaPoly-1305` encryption using an all-zero nonce).
      - MUST set `enctlv` to a valid `encmsg_tlvs` which meets the requirements
        of the `onionmsg_payload` when used by the recipient.
      - MAY use `self_id` to contain a secret so it can recognize use of this `reply_path`.
  - otherwise:
    - MUST NOT set `reply_path`.

The reader:
- if it is not the final node according to the onion encryption:
  - if `enctlv` is not present, or does not decrypt with the shared secret from the given `blinding` parameter:
    - MUST drop the message.
  - if the `enctlv` is not a valid TLV or does not contain `next_node_id`:
    - MUST drop the message.
  - if the `enctlv` contains `self_id`:
    - MUST drop the message.
  - otherwise:
    - MUST ignore `padding`, if any.
    - SHOULD forward the message using `onion_message` to the next peer indicated by `next_node_id`.
    - if it forwards the message:
      - if `blinding` is specified in the `enctlv`:

>          - MUST pass that as `blinding` in the `onion_message`

Reviewer 5: Why not always use the derived `blinding`? What is the use-case where we need to override the default `blinding`?

Author: If I give you a blinded path from Alice, you need to make an onion which goes to Alice then uses the blinded path.  This requires blinding override.

Reviewer: Can you detail that @rustyrussell? I don't understand it either.
If for example Carol creates a blinded path `Alice -> Bob -> Carol`.
Carol provides a list of `encrypted_recipient_data` and a first `blinding` point that must be sent to Alice.
But that `blinding` cannot be inside the `encrypted_recipient_data` as it's required to actually decrypt it...?
It seems to me that you just need the `blinding` tlv fields that is part of the `onion_message` for that.

Reviewer 5: I receive the following blinded route built by Carol: `(blinding0, Alice, (BlindedAlice, BlindedAliceData) -> (BlindedBob, BlindedBobData) -> (BlindedCarol, BlindedCarolData))` and I want use it by adding Dave in front.

I want to create a route `(blinding1, Dave, (BlindedDave, BlindedDaveData) -> (BlindedAlice, BlindedAliceData) -> (BlindedBob, BlindedBobData) -> (BlindedCarol, BlindedCarolData))`. For that I generate randomly `blinding1` and use it to blind `(Dave, DaveData)`, I then obtain `(BlindedDave, BlindedDaveData)` and a derived `blinding1Next`. However `blinding1Next` is not equal to `blinding0`, so I need to specify `blinding0` as an override for the rest of the route. I pass this `blinding0` to Dave, inside `DaveData` and Dave MUST pass this `blinding0` as `blinding` in the `onion_message` it sends to Alice.

Reviewer: If you're concatenating two blinded routes I can see why you would need that, but is there a use-case for that though?
In all scenarios I can think of you would use a normal route through Dave, the blinded part would only start at Alice.
The recipient is the one that wants to hide, why would the sender would use a blinded route for the part that he chooses?

Reviewer 5: There is no such thing as a normal route. Everything is blinded.

Reviewer: Right, that makes sense! It's worth explaining a bit more in the PR then, the rationale section should make this obvious, since route blinding is designed to prepend a normal route to a blinded one (at least for payment scenarios).

      - otherwise:
         - MUST pass `blinding` derived as in [Route Blinding][route-blinding] (i.e. `E(i+1) = H(E(i) || ss(i)) * E(i)`).

- otherwise:
  - if this was sent via a `reply_path` the reader previously supplied:
    - if it is not a reply to the onion message sent with that `reply_path`
      - MUST ignore the onion message.
  - otherwise:
    - if this is a reply to an onion message which was sent with a `reply_path`:
      - MUST respond (or not respond) exactly as if it did not send the initial onion message.
  - if it wants to send a reply:

>     - MUST create an onion encoding using `reply_path`.

Reviewer 5: This needs to be explained with more details.

Author: Yes...

  - if the final node is permitted to reply:
    - MUST set `reply_path` `blinding` to the initial blinding factor for the `next_node_id`
    - MUST set `reply_path` `first_node_id` to the unblinded node id of the first node in the reply path.
    - For every `reply_path` `path`:
      - MUST set `node_id` to the blinded node id to encrypt the onion hop for.
      - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
        `ChaChaPoly-1305` encryption using an all-zero nonce).
      - MUST set `enctlv` to a valid `encmsg_tlvs` which meets the requirements
        of the `onionmsg_payload` when used by the recipient.
      - MAY use `path_id` to contain a secret so it can recognize use of this `reply_path`.
  - otherwise:
    - MUST NOT set `reply_path`.

The reader:
- if it is not the final node according to the onion encryption:
  - if `enctlv` is not present, or does not decrypt with the shared secret from the given `blinding` parameter:
    - MUST drop the message.
  - if the `enctlv` is not a valid TLV or does not contain `next_node_id`:
    - MUST drop the message.
  - if the `enctlv` contains `path_id`:
    - MUST drop the message.
  - otherwise:

>     - MUST ignore `padding`, if any.

Author: You'd *think* so, but Id' rather say *exactly* what to do with each field, so...

intermediary nodes could simply claim the remaining ones.

### Onion Messages

Onion messages have an onion with an alternate `hop_payload`
format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
is no legacy format, thus a `bigsize` of 0 means no payload.

1. tlvs: `onionmsg_payload`
2. types:
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]
    1. type: 6 (`next_short_channel_id`)
    2. data:
        * [`short_channel_id`:`short_channel_id`]
    1. type: 8 (`reply_path`)
    2. data:
        * [`point`:`blinding`]
        * [`...*onionmsg_path`:`path`]
    1. type: 10 (`enctlv`)
    2. data:
        * [`...*byte`:`enctlv`]
    1. type: 12 (`blinding`)
    2. data:
        * [`point`:`blinding`]

1. tlvs: `encmsg_tlvs`
2. types:
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]
    1. type: 6 (`next_short_channel_id`)
    2. data:
        * [`short_channel_id`:`short_channel_id`]

1. subtype: `onionmsg_path`
2. data:
    * [`point`:`node_id`]
    * [`u16`:`enclen`]
    * [`enclen*byte`:`enctlv`]

#### Requirements

The writer:
- For the non-final nodes' `onionmsg_payload`:

>   - MUST include exactly one of `next_short_channel_id`, `next_node_id`

Reviewer 3: Why not require they be encrypted? It seems like a bunch of extra code to handle being optionally-encrypted when we could just require it. I think there should at least be a rationale section that describes why someone would want to send an unencrypted next-hop.

Author: Agreed, this is now fixed.

invoice timeout), however if it were to fulfill only some of them,
intermediary nodes could simply claim the remaining ones.

### Route Blinding

Nodes receiving onion packets may hide their identity from senders by
"blinding" an arbitrary amount of hops at the end of an onion path.

When using route blinding, nodes find a route to themselves from a given
"introduction node". They then use ECDH with each node in that route to create

> a "blinded" node ID and an encrypted blob (`encrypted_data`) for each one of

Reviewer 6: See comments below for more context but I don't understand the extra use coming from the additional complexity of the blinded node id. Is that just a name for the variable to explain the formulas?

The actual encrypted node_ids need to be stored in `next_node_id` anyway and the blinded id can always be computed from the ephemeral shared secret and the actual node id (as currently specified in line 473 to 475) but not the other way around. So the blinded node ids seem to be useless to me.

the blinded nodes.

They communicate this blinded route and the encrypted blobs to the sender.
The sender finds a route to the introduction node and extends it with the
blinded route provided by the recipient. The sender includes the encrypted
blobs in the corresponding onion payloads: they allow nodes in the blinded
part of the route to "unblind" the next node and correctly forward the packet.

The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
may contain the following TLV fields:

1. `tlv_stream`: `encrypted_data_tlv`
2. types:
    1. type: 1 (`padding`)
    2. data:
        * [`...*byte`:`padding`]

>     1. type: 2 (`short_channel_id`)

Reviewer 7: As discussed on the BOLT 12 Telegram, it may make sense to get rid of this field since `next_node_id` is currently mandatory, thus making the scid redundant. In addition, IIUC Acinq will currently error if the `next_node_id` field is not present (cc @t-bast).

Reviewer: I don't think we should remove it. For payments we still want to use the `short_channel_id` to identify the next node as it uses less space in the onion. For onion messages we're only using `next_node_id`, but for blinded payments we use `short_channel_id`.

Reviewer: We can calculate that, but I'd rather just keep this option of using less space, it can always be useful (e.g. to add more dummy hops at the end of a route). This is what we use for normal payments relay, so it shouldn't require implementations to add any new code, so I don't see why we'd want to remove it?

Reviewer 7: > I don't think we should remove it. For payments we still want to use the `short_channel_id` to identify the next node as it uses less space in the onion. For onion messages we're only using `next_node_id`, but for blinded payments we use `short_channel_id`.

That seems fair. Based on that, I'd prefer if we made `next_node_id` explicitly mandatory for onion messages and disallow `scid`, and the reverse requirements for payment relay. (I believe that's what eclair does now, at least the `next_node_id` part?)

    2. data:
        * [`short_channel_id`:`short_channel_id`]
    1. type: 4 (`next_node_id`)
    2. data:
        * [`point`:`node_id`]
    1. type: 6 (`path_id`)
    2. data:
        * [`...*byte`:`data`]
    1. type: 8 (`next_blinding_override`)
    2. data:
        * [`point`:`blinding`]

#### Requirements

> A recipient N(r) creating a blinded route `N(0) -> N(1) -> ... -> N(r)` to itself:

Reviewer 6: Just to be clear: `N(0)` is the "introduction node"? Maybe we could add this similar to the other explanations in the following lines.

- MUST create a blinded node ID `B(i)` for each node using the following algorithm:
  - `e(0) <- {0;1}^256`
  - `E(0) = e(0) * G`
  - For every node `N(i)` in the route:
    - let `P(i) = k(i) * G` be `N(i)`'s `node_id` (`k(i)` is `N(i)`'s private key)
    - `ss(i) = SHA256(e(i) * P(i)) = SHA256(k(i) * E(i))` (ECDH shared secret known only by `N(r)` and `N(i)`)
    - `B(i) = HMAC256("blinded_node_id", ss(i)) * P(i)` (blinded `node_id` for `N(i)`, private key known only by `N(i)`)
    - `rho(i) = HMAC256("rho", ss(i))`
    - `e(i+1) = SHA256(E(i) || ss(i)) * e(i)` (ephemeral private key, only known by `N(r)`)

>     - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)` (NB: `N(i)` MUST NOT learn `e(i)`)

Reviewer 6: Since we made the ECDH explicit in line 430 we could also here help a bit by writing:

```
E(i+1) = e(i+1)*G 
       = SHA256(E(i) || ss(i)) * e(i) * G
       = SHA256(E(i) || ss(i)) * E(i)
```

with the comment from line 427 everyone can see that the math is correctly

> - SHOULD create an `encrypted_data` for each blinded node `B(i)`

Reviewer 6: Just to be clear `next_node_id` should be `N(i+1)` and not `B(i+1)`. It is strange to me to call the nodes on the blinded path `N(0)--> N(r)` now `B(i)`.

- If it creates `encrypted_data`:
  - MUST encrypt it with ChaCha20-Poly1305 using the `rho(i)` key and an all-zero nonce
  - MAY store data in the `path_id` of the payload to itself to verify that

>     the route is used in the right context

Reviewer 6: How does that verification work? What is meant with "right context"? I see that there was already rational explaining this so maybe here it is a bit redundant?

  - MAY add padding data to ensure all `encrypted_data` have the same length

> - MUST communicate the blinded node IDs `B(i)` and `encrypted_data(i)` to the sender

Reviewer 6: Why?
Given `N(i)` which you require the node can also compute the `B(i)` with the help of `E(i)`

- MUST communicate the real node ID of the introduction point `N(0)` to the sender
- MUST communicate the first ephemeral key `E(0)` to the sender

The sender:

- MUST provide the `encrypted_data` field for each node in the blinded route
  in their respective onion payloads
- MUST communicate `E(0)` to the introduction point

The introduction point:

- MUST compute:
  - `ss(0) = SHA256(k(0) * E(0))` (standard ECDH)
  - `rho(0) = HMAC256("rho", ss(0))`
  - `E(1) = SHA256(E(0) || ss(0)) * E(0)`
- If an `encrypted_data` field is provided:
  - MUST decrypt it using `rho(0)` and `E(0)`
  - MUST use the decrypted fields to locate the next node
- If `encrypted_data` contains a `next_blinding_override`:
  - MUST use it as the next blinding point
- Otherise:

>   - MUST use `E(1)` as the next blinding point

Reviewer 6: While `E(i)` makes a lot of sense the notation blinding point is actually a bit confusing as I initially thought `B(1)` which as the blinded node_id is also a point. But I guess this all boils down to the general gist / suggestions to remove the notation of a blinded node_id `B(i)`

- MUST forward the onion and communicate the next blinding point to the next node

An intermediate node in the blinded route:

- MUST compute:
  - `ss(i) = SHA256(k(i) * E(i))` (standard ECDH)

>   - `b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)`

Reviewer 6: If `B(i)` was actually explicitly communicated to `N(i)` the node could verify that `B(i) = b(i)*G` but other than that I do not see any use of the `B(i)`

  - `rho(r) = HMAC256("rho", ss(r))`
- If an `encrypted_data` field is provided:
  - MUST decrypt it using `rho(r)` and `E(r)`
- MUST ignore the message if `E(r)` or `path_id` does not match the
  blinded route it created

#### Rationale

Route blinding is a lightweight technique to provide recipient anonymity.
It's more flexible than rendezvous because it lets senders choose what data
they include in each node's onion payload and retry with updated values if
the route fails.

Use-cases where route blinding is useful include:

- Recipient anonymity when receiving payments
- Using unannounced channels in invoices without revealing them
- Forcing a message to go through a specific set of intermediaries that can witness it
- Providing anonymous reply paths for onion messages (offers, stuckless payments, etc)

Each node in the blinded route needs to receive `E(i)` to be able to decrypt
the onion and the `encrypted_data`. Protocols that use route blinding must
specify how this value is propagated between nodes.

When concatenating two blinded routes generated by different nodes, the
last node of the first route needs to know the first `blinding` of the second

> route: the `next_blinding_override` field must be used to transmit this data.

Reviewer 6: Given I understand this correctly I think this is a neat trick to do the ephemeral key switch!

> The final recipient must verify that the blinded route is used in the right

Reviewer 6: why is "must" not written capitalized here?

context (e.g. for a specific payment). It can do so by storing `E(r)` and the
context (e.g. a `payment_hash`), and verifying that they match when receiving
the onion. Otherwise, to avoid additional storage cost, it can put some context
information in the `path_id` field (e.g. a `payment_hash`) and verify that when
receiving the onion.

The `padding` field can be used to ensure that all `encrypted_data` have the
same length. It's particularly useful when adding dummy hops at the end of a
blinded route, to prevent the sender from figuring out which node is the final
recipient.

### Onion Message Payload Format

Onion messages have an onion with an alternate `hop_payload`
format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
is no legacy format, thus a `bigsize` of 0 means no payload.

An intermediate node expects an `encrypted_data_tlv` which it can
decrypt using the `blinding` which it is handed along with the onion
message.

Field numbers 64 and above are reserved for payloads for the final
hop.

1. `tlv_stream`: `onionmsg_payload`
2. types:

>     1. type: 2 (`reply_path`)

Reviewer 8: How is this secure from spoofing the `reply_path`? 

As the originator from the Onion Message is not known couldn't one basically create a `reply_path` to any node in the network and create potential DoS attacks (especially if rate limiting of onion messages is being introduced)

Reviewer 5: You would have to send one message to a random node so that your target is spammed with one message. How is that better than just sending the message to your target directly? You can indeed do that but it doesn't amplify the number of messages.

For a DoS attack you would more likely want to have as many hops as possible to amplify the attack but this reflection doesn't give you anything interesting.

Reviewer 8: True, unless I find a way in which the node will send the reply via the `reply_path` several times because it thinks the reply has not been received as we use an unreliable protocol (this could eventually happen from application land)

Reviewer 8: I suppose this will work. Though it seems still a bit risky

    2. data:

>         * [`point`:`first_node_id`]

Reviewer 7: Discussed a while ago on [telegram](https://t.me/bolt12org/1687) -- do we want to get rid of this field and put the unblinded `first_node_id` in the blinded path's first hop? I believe offers encoding does this currently

Author: I thought @thomash-acinq pointed out that you can't do this without knowing the blinding secret :(  I think that may be correct...

Reviewer 7: Ah! That checks out to me, thanks

blinded route, to prevent the sender from figuring out which node is the final
recipient.

### Onion Message Payload Format

Onion messages have an onion with an alternate `hop_payload`
format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
is no legacy format, thus a `bigsize` of 0 means no payload.

An intermediate node expects an `encrypted_data_tlv` which it can
decrypt using the `blinding` which it is handed along with the onion
message.

Field numbers 64 and above are reserved for payloads for the final
hop.

1. `tlv_stream`: `onionmsg_payload`
2. types:
    1. type: 2 (`reply_path`)
    2. data:
        * [`point`:`first_node_id`]
        * [`point`:`blinding`]
        * [`...*onionmsg_path`:`path`]

>     1. type: 4 (`encrypted_data_tlv`)

Reviewer 7: I like `encrypted_tlv_stream` or `encrypted_tlvs` but perhaps overly explicit lol

>     1. type: 4 (`encrypted_data_tlv`)

Author: The route blinding calls this `encrypted_data`, which gets decrypted into a `encrypted_data_tlv`.  Lets' be consistent: if we don't like it we can fix all of it together...

>     1. type: 10 (`encrypted_data_tlv`)

Reviewer: We're currently using 4 in eclair:

```suggestion
    1. type: 4 (`encrypted_data_tlv`)
```

I don't see a good reason to leave a gap here, it's a bit confusing because it looks like the tlv namespace is shared between `onionmsg_payload` and `encrypted_data`, which isn't the case.

Author: Yep, I came here to change that :)

    2. data:
        * [`...*byte`:`encrypted_data_tlv`]

1. subtype: `onionmsg_path`
2. data:
    * [`point`:`node_id`]
    * [`u16`:`enclen`]
    * [`enclen*byte`:`encrypted_recipient_data`]

#### Requirements

The writer:
- For the non-final nodes' `onionmsg_payload`:

>   - MUST set `enctlv` to a valid `encrypted_data_tlv` stream containing `next_node_id`.

Reviewer 4: `enctlv` has been renamed in the format above but not in the requirements section

Author: Oops, fixed!

[fips198]: http://csrc.nist.gov/publications/fips/fips198-1/FIPS-198-1_final.pdf
[sec2]: http://www.secg.org/sec2-v2.pdf
[rfc8439]: https://tools.ietf.org/html/rfc8439

> [routeblinding]: https://github.com/lightningnetwork/lightning-rfc/blob/route-blinding/proposals/route-blinding.md

Reviewer 3: This needs to be in the repo if we want to reference it, presumably in this PR. I have a bunch of comments for this file to make clearer what's going on, but can't easily leave comments, and we shouldn't be referencing another branch of the same repo.

  - MAY use the data specified in the various failure types for debugging
  purposes.

# Onion Messages

Onion messages allow peers to use existing connections to query for
invoices (see [BOLT 12](12-offer-encoding.md)).  Like gossip messages,
they are not associated with a particular local channel.  Like HTLCs,
they use [onion messages](#onion-message-payload-format) protocol for
end-to-end encryption.

Onion messages are unreliable: in particular, they are designed to
be cheap to process and require no storage to forward.  As a result,
there is no error returned from intermediary nodes.

For consistency, all onion messages use [Route Blinding](#route-blinding).

## The `onion_message` Message

1. type: 513 (`onion_message`) (`option_onion_messages`)
2. data:
    * [`point`:`blinding`]
    * [`u16`:`len`]

>     * [`len*byte`:`onionmsg`]

Reviewer 9: can we call this `onionmsg_payloads` and refer to it in the section mentioning `hop_payload` to make the linkage explicit?

Author: I called them both "onion_message_packet" to be clear...

## Requirements

The writer:
- MUST populate the per-hop payloads as described in [Onion Message Payload Format](onion-message-payload-format).
- SHOULD retry via a different route if it expects a response and
  doesn't receive one after a reasonable period.
- SHOULD set `len` to 1366 or 32834.

The reader:
- MUST handle the per-hop payloads as described in [Onion Message Payload Format](onion-message-payload-format).

> - SHOULD accept onion messages from peers without an established channel.

Reviewer 6: I assume this is with Offers in mind but it is strange to me. Given the ongoing discussion about spam prevention and rate-limiting it seems reasonable to require at least a payment channel before accepting and forwarding onion messages

Author: You don't have to forward them, or you can ratelimit them, but you should definitely accept them.  Like you do pings...

Author: Noted in rationale.

as the `hop_payloads` format, except there is no "legacy" length: a 0
`length` would mean an empty `onionmsg_payload`.

Onion messages are unreliable: in particular, they are designed to
be cheap to process and require no storage to forward.  As a result,
there is no error returned from intermediary nodes.

For consistency, all onion messages use [Route Blinding](#route-blinding).

## The `onion_message` Message

1. type: 513 (`onion_message`) (`option_onion_messages`)
2. data:
    * [`point`:`blinding`]
    * [`u16`:`len`]
    * [`len*byte`:`onion_message_packet`]

1. type: `onion_message_packet`
2. data:
   * [`byte`:`version`]
   * [`point`:`public_key`]
   * [`...*byte`:`onionmsg_payloads`]
   * [`32*byte`:`hmac`]

1. type: `onionmsg_payloads`
2. data:
   * [`bigsize`:`length`]
   * [`length*u8`:`onionmsg_tlv`]
   * [`32*byte`:`hmac`]
   * ...
   * `filler`

The `onionmsg_tlv` itself is a TLV: an intermediate node expects an
`encrypted_tlv_stream` which it can decrypt into an `encrypted_data_tlv`
using the `blinding` which it is handed along with the onion message.

Field numbers 64 and above are reserved for payloads for the final

> hop.

Author: No, there is no such *requirement*, and I don't think we need to box ourselves in on that.  Just a convention.

I'll add a note.

    - MUST send the reply via `onion_message` to the node indicated by
        the `first_node_id`, using `reply_path` `blinding` to send
        along `reply_path` `path`.

#### Rationale

Care must be taken that replies are only accepted using the exact
reply_path given, otherwise probing is possible.  That means checking
both ways: non-replies don't use the reply path, and replies always
use the reply path.

The requirement to discard messages with `onionmsg_tlv` fields which
are not strictly required ensures consistency between current and
future implementations.  Even odd fields can be a problem since they
are parsed (and thus may be rejected!) by nodes which understand them,
and ignored by those which don't.

All onion messages are blinded, even though this overhead is not
always necessary (33 bytes here, the 16-byte MAC for each encrypted_data_tlv in
the onion).  This blinding allows nodes to use a path provided by
others without knowing its contents.  Using it universally simplifies
implementations a little, and makes it more difficult to distinguish
onion messages.

`len` allows larger messages to be sent than the standard 1300 bytes

> allowed for an HTLC onion, but this should be used sparingly as it is
> reduces anonymity set, hence the recommendation that it either look

Reviewer 10: nit:
```suggestion
allowed for an HTLC onion, but this should be used sparingly as it
reduces the anonymity set, hence the recommendation that it either looks
```

Author: Applied, thanks!

  - MAY use the data specified in the various failure types for debugging
  purposes.

# Onion Messages

Onion messages allow peers to use existing connections to query for
invoices (see [BOLT 12](12-offer-encoding.md)).  Like gossip messages,
they are not associated with a particular local channel.  Like HTLCs,
they use [onion messages](#onion-messages) protocol for
end-to-end encryption.

Onion messages use the same form as HTLC `onion_packet`, with a
slightly more flexible format: instead of 1300 byte payloads, the
payload length is implied by the total length (minus 66 bytes for the
header and trailing bytes).  The `onionmsg_payloads` themselves are the same
as the `hop_payloads` format, except there is no "legacy" length: a 0
`length` would mean an empty `onionmsg_payload`.

> Onion messages are unreliable: in particular, they are designed to

Reviewer 11: Since OMs are unreliable and a sender might have to retry multiple times (maybe even try numerous paths simultaneously to gain more reliability), does it make sense to add an idempotency key to onion messages?

    1. type: 4 (`encrypted_recipient_data`)
    2. data:
        * [`...*byte`:`encrypted_recipient_data`]

1. subtype: `blinded_path`
2. data:
   * [`point`:`first_node_id`]
   * [`point`:`blinding`]
   * [`byte`:`num_hops`]
   * [`num_hops*onionmsg_hop`:`path`]

1. subtype: `onionmsg_hop`
2. data:
    * [`point`:`blinded_node_id`]
    * [`u16`:`enclen`]
    * [`enclen*byte`:`encrypted_recipient_data`]

#### Requirements

The creator of `encrypted_recipient_data` (usually, the recipient of the onion):

  - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
  - MUST NOT include `short_channel_id`, `payment_relay` or `payment_constraints` in any `encrypted_data_tlv`
  - MUST include `encrypted_data_tlv.next_node_id` for each non-final node.
  - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).

The writer:

- MUST set the `onion_message_packet` `version` to 0.
- MUST construct the `onion_message_packet` `onionmsg_payloads` as detailed above using Sphinx.
- MUST NOT use any `associated_data` in the Sphinx construcion.

> - SHOULD set `onion_message_packet` `len` to 1366 or 32834.

Reviewer 11: Does this mean that an OM can travel far more than 20 hops?

Reviewer 12: Potentially, yes. With varonions the limitation to fixed 20 hops went away anyway, so even the smaller 1366 byte onion could potentially contain more hops. It all deppends on the payload you want to deliver at each hop.

Reviewer 11: Thank you for the answer. So if I want to calculate the maximum number of hops I can go through, does this look right?
32834 / 66 =~ 497 hops

using the `blinding` which it is handed along with the onion message.

Field numbers 64 and above are reserved for payloads for the final
hop, though these are not explicitly refused by non-final hops (unless
even, of course!).

1. `tlv_stream`: `onionmsg_tlv`
2. types:
    1. type: 2 (`reply_path`)
    2. data:
        * [`blinded_path`:`path`]
    1. type: 4 (`encrypted_recipient_data`)
    2. data:
        * [`...*byte`:`encrypted_recipient_data`]

1. subtype: `blinded_path`
2. data:
   * [`point`:`first_node_id`]
   * [`point`:`blinding`]
   * [`byte`:`num_hops`]
   * [`num_hops*onionmsg_hop`:`path`]

1. subtype: `onionmsg_hop`
2. data:
    * [`point`:`blinded_node_id`]
    * [`u16`:`enclen`]
    * [`enclen*byte`:`encrypted_recipient_data`]

#### Requirements

The creator of `encrypted_recipient_data` (usually, the recipient of the onion):

  - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
  - MUST NOT include `short_channel_id`, `payment_relay` or `payment_constraints` in any `encrypted_data_tlv`
  - MUST include `encrypted_data_tlv.next_node_id` for each non-final node.

>   - MUST NOT include any other fields in `encrypted_data_tlv` for any non-final node.

Author: Yes, this is dumb: we don't need to worry about the creator of the path de-anonymizing.   Removed.

This was presumably supposed to be a requirement on onionmsg_tlv: We already say that non-final nodes should drop onionmsg_tlv which contains anything other than `encrypted_recipient_data`, but we don't say you shouldn't put anything else in onionmsg_tlv:

```diff
diff --git a/04-onion-routing.md b/04-onion-routing.md
index cebd8a9..0e95b39 100644
--- a/04-onion-routing.md
+++ b/04-onion-routing.md
@@ -1488,7 +1488,6 @@ The creator of `encrypted_recipient_data` (usually, the recipient of the onion):
   - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
   - MUST NOT include `short_channel_id`, `payment_relay` or `payment_constraints` in any `encrypted_data_tlv`
   - MUST include `encrypted_data_tlv.next_node_id` for each non-final node.
-  - MUST NOT include any other fields in `encrypted_data_tlv` for any non-final node.
   - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
 
 The writer:
@@ -1499,7 +1498,7 @@ The writer:
 - SHOULD set `onion_message_packet` `len` to 1366 or 32834.
 - SHOULD retry via a different path if it expects a response and doesn't receive one after a reasonable period.
 - For the non-final nodes' `onionmsg_tlv`:
-  - MUST NOT set `reply_path` 
+  - MUST NOT set fields other than `encrypted_recipient_data`.
 - For the final node's `onionmsg_tlv`:
   - if the final node is permitted to reply:
     - MUST set `reply_path` `blinding` to the initial blinding factor for the `first_node_id`
```

The writer:

- MUST set the `onion_message_packet` `version` to 0.
- MUST construct the `onion_message_packet` `onionmsg_payloads` as detailed above using Sphinx.
- MUST NOT use any `associated_data` in the Sphinx construcion.
- SHOULD set `onion_message_packet` `len` to 1366 or 32834.
- SHOULD retry via a different path if it expects a response and
  doesn't receive one after a reasonable period.
- For the non-final nodes' `onionmsg_tlv`:
  - MUST NOT set `reply_path` 
- For the final node's `onionmsg_tlv`:
  - if the final node is permitted to reply:
    - MUST set `reply_path` `blinding` to the initial blinding factor for the `first_node_id`
    - MUST set `reply_path` `first_node_id` to the unblinded node id of the first node in the reply path.
    - For every `reply_path` `path`:
      - MUST set `blinded_node_id` to the blinded node id to encrypt the onion hop for.
      - MUST set `encrypted_recipient_data` to a valid encrypted `encrypted_data_tlv` stream which meets the requirements of the `onionmsg_tlv` when used by the recipient.
      - MAY use `path_id` to contain a secret so it can recognize use of this `reply_path`.
  - otherwise:
    - MUST NOT set `reply_path`.

> - SHOULD retry via a different route if it expects a response and doesn't receive one after a reasonable period.

Reviewer 13: ..and lines 1500 and 1501 should be combined on one line

The reader:

- SHOULD accept onion messages from peers without an established channel.
- MAY rate-limit messages by dropping them.
- MUST read the `encrypted_recipient_data` using `blinding` as required in [Route Blinding](#route-blinding).
  - MUST ignore the message if that considers the message invalid.
- if `encrypted_data_tlv` contains `allowed_features`:
  - MUST ignore the message if:
    - `encrypted_data_tlv.allowed_features.features` contains an unknown feature bit (even if it is odd).

>     - the payment uses a feature not included in `encrypted_data_tlv.allowed_features.features`.

Reviewer: ```suggestion
    - the message uses a feature not included in `encrypted_data_tlv.allowed_features.features`.
```

- if it is not the final node according to the onion encryption:
  - if the `onionmsg_tlv` contains other tlv fields than `encrypted_recipient_data`:
    - MUST ignore the message.
  - if the `encrypted_data_tlv` contains `path_id`:
    - MUST ignore the message.
  - otherwise:
    - SHOULD forward the message using `onion_message` to the next peer indicated by `next_node_id`.
    - if it forwards the message:
      - MUST set `blinding` in the forwarded `onion_message` to the next blinding as calculated in [Route Blinding](#route-blinding).
- otherwise (it is the final node):
  - if `path_id` is set and corresponds to a path the reader has previously published in a `reply_path`:
    - if the onion message is not a reply to that previous onion:
      - MUST ignore the onion message
  - otherwise (unknown or unset `path_id`):
    - if the onion message is a reply to an onion message which contained a `path_id`:
      - MUST respond (or not respond) exactly as if it did not send the initial onion message.
  - if the `onionmsg_tlv` contains other tlv fields than `encrypted_recipient_data` and `reply_path`:

>     - MUST ignore the message.

Reviewer 5: Why? We actually want to use other fields, without them onion messages are not very useful.
It should be MUST ignore fields that are not expected.

Author: Indeed!

I think we should ignore the message if more than one payload field is set?  If one implementation/release version accepts both invreq and inv, and another doesn't, it's a trivial way to detect?

Author: ```diff
diff --git a/04-onion-routing.md b/04-onion-routing.md
index c18a004..cebd8a9 100644
--- a/04-onion-routing.md
+++ b/04-onion-routing.md
@@ -1538,7 +1538,7 @@ The reader:
   - otherwise (unknown or unset `path_id`):
     - if the onion message is a reply to an onion message which contained a `path_id`:
       - MUST respond (or not respond) exactly as if it did not send the initial onion message.
-  - if the `onionmsg_tlv` contains other tlv fields than `encrypted_recipient_data` and `reply_path`:
+  - if the `onionmsg_tlv` contains more than one payload field:
     - MUST ignore the message.
   - if it wants to send a reply:
     - MUST create an onion message using `reply_path`.
```

    - MUST send the reply via `onion_message` to the node indicated by
      the `first_node_id`, using `reply_path` `blinding` to send
      along `reply_path` `path`.

#### Rationale

Care must be taken that replies are only accepted using the exact
reply_path given, otherwise probing is possible.  That means checking
both ways: non-replies don't use the reply path, and replies always
use the reply path.

The requirement to discard messages with `onionmsg_tlv` fields which
are not strictly required ensures consistency between current and
future implementations.  Even odd fields can be a problem since they
are parsed (and thus may be rejected!) by nodes which understand them,
and ignored by those which don't.

All onion messages are blinded, even though this overhead is not
always necessary (33 bytes here, the 16-byte MAC for each encrypted_data_tlv in
the onion).  This blinding allows nodes to use a path provided by
others without knowing its contents.  Using it universally simplifies
implementations a little, and makes it more difficult to distinguish
onion messages.

`len` allows larger messages to be sent than the standard 1300 bytes

> allowed for an HTLC onion, but this should be used sparingly as it is
> reduces the anonymity set, hence the recommendation that it either looks

Reviewer: ```suggestion
allowed for an HTLC onion, but this should be used sparingly as it
reduces the anonymity set, hence the recommendation that it either looks
```
